File: PartialPreprocessor/src/main/java/org/anarres/cpp/MacroTokenSource.java
/*
 * Anarres C Preprocessor
 * Copyright (c) 2007-2008, Shevek
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 */

package org.anarres.cpp;

import static org.anarres.cpp.Token.CCOMMENT;
import static org.anarres.cpp.Token.CPPCOMMENT;
import static org.anarres.cpp.Token.EOF;
import static org.anarres.cpp.Token.M_ARG;
import static org.anarres.cpp.Token.M_PASTE;
import static org.anarres.cpp.Token.M_STRING;
import static org.anarres.cpp.Token.STRING;

import java.io.IOException;
import java.util.Iterator;
import java.util.List;

/* This source should always be active, since we don't expand macros
 * in any inactive context. */
/* pp */class MacroTokenSource extends Source {
	private final MacroData macro;
	private Iterator<Token> tokenIter; /* Pointer into the macro. */
	private List<Argument> args; /* { unexpanded, expanded } */
	private Iterator<Token> arg; /* "current expansion" */
	private final String macroName;
	private boolean gnuCExtensions;

	/* pp */MacroTokenSource(String macroName, MacroData m, List<Argument> args, boolean gnuCExtensions) {
		this.macroName = macroName;
		this.macro = m;
		this.tokenIter = m.getTokens().iterator();
		this.args = args;
		this.arg = null;
		this.gnuCExtensions = gnuCExtensions;
	}

	@Override
	/* pp */boolean mayExpand(String macroName) {
		// /*
		// * When we are expanding an arg, 'this' macro is not being expanded,
		// and
		// * thus we may re-expand it.
		// */
		if (macroName.equals(this.macroName))
			return false;
		// if (/* XXX this.arg == null && */ this.macro == m)
		// return true;
		return super.mayExpand(macroName);
	}

	/* XXX Called from Preprocessor [ugly]. */
	/* pp */static void escape(StringBuilder buf, CharSequence cs) {
		for (int i = 0; i < cs.length(); i++) {
			char c = cs.charAt(i);
			switch (c) {
			case '\\':
				buf.append("\\\\");
				break;
			case '"':
				buf.append("\\\"");
				break;
			case '\n':
				buf.append("\\n");
				break;
			case '\r':
				buf.append("\\r");
				break;
			default:
				buf.append(c);
			}
		}
	}

	private void concat(StringBuilder buf, Argument arg, boolean queuedComma) {
		if (queuedComma) {
			if (!arg.isOmittedArg() || !gnuCExtensions) {
				//Output the comma that we didn't output previously.
				buf.append(",");
			} else {
				//Swallow the semicolon, as prescribed by:
				// http://gcc.gnu.org/onlinedocs/cpp/Variadic-Macros.html
				assert arg.isEmpty();
				return;
			}
		}
		Iterator<Token> it = arg.iterator();
		while (it.hasNext()) {
			Token tok = it.next();
			buf.append(tok.getText());
		}
	}

	private Token stringify(Token pos, Argument arg) {
		StringBuilder buf = new StringBuilder();
		concat(buf, arg, false);
		// System.out.println("Concat: " + arg + " -> " + buf);
		StringBuilder str = new StringBuilder("\"");
		escape(str, buf);
		str.append("\"");
		// System.out.println("Escape: " + buf + " -> " + str);
		return new SimpleToken(STRING, pos.getLine(), pos.getColumn(),
				str.toString(), buf.toString(), this);
	}

	/*
	 * At this point, we have consumed the first M_PASTE.
	 * 
	 * @see Macro#addPaste(Token)
	 */
	private void paste(Token ptok) throws IOException, LexerException {
		StringBuilder buf = new StringBuilder();
		boolean queuedComma = false;
		/*
		 * We know here that arg is null or expired, since we cannot paste an
		 * expanded arg.
		 */

		int count = 2;
		for (int i = 0; i < count; i++) {
			if (!tokenIter.hasNext()) {
				/* XXX This one really should throw. */
				error(ptok.getLine(), ptok.getColumn(),
						"Paste at end of expansion");
				buf.append(' ').append(ptok.getText());
				break;
			}
			Token tok = tokenIter.next();
			if (queuedComma && tok.getType() != M_ARG) {
				buf.append(",");
				queuedComma = false;
			}
			// System.out.println("Paste " + tok);
			switch (tok.getType()) {
			case M_PASTE:
				/*
				 * One extra to paste, plus one because the paste token didn't
				 * count.
				 */
				count += 2;
				ptok = tok;
				break;
			case M_ARG:
				int idx = ((Integer) tok.getValue()).intValue();
				concat(buf, args.get(idx), queuedComma);
				break;
			/* XXX Test this. */
			case CCOMMENT:
			case CPPCOMMENT:
				break;
			case ',':
				assert ",".equals(tok.getText());
				queuedComma = true;
				break;
			default:
				buf.append(tok.getText());
				break;
			}
		}

		/* Push and re-lex. */
		/*
		 * StringBuilder src = new StringBuilder(); escape(src, buf);
		 * StringLexerSource sl = new StringLexerSource(src.toString());
		 */
		StringLexerSource sl = new StringLexerSource(buf.toString());

		/* XXX Check that concatenation produces a valid token. */

		arg = new SourceIterator(sl);
	}

	public Token token() throws IOException, LexerException {
		Token tok = _token();
		if (tok.getText().equals(macroName))
			tok.setNoFurtherExpansion();
		return tok;
	}

	public Token _token() throws IOException, LexerException {
		for (;;) {
			/* Deal with lexed tokens first. */

			if (arg != null) {
				if (arg.hasNext()) {
					Token tok = arg.next();
					/* XXX PASTE -> INVALID. */
					assert tok.getType() != M_PASTE : "Unexpected paste token";
					return tok;
				}
				arg = null;
			}

			if (!tokenIter.hasNext())
				return new SimpleToken(EOF, -1, -1, "", this); /* End of macro. */
			Token tok = tokenIter.next();
			int idx;
			switch (tok.getType()) {
			case M_STRING:
				/* Use the nonexpanded arg. */
				idx = ((Integer) tok.getValue()).intValue();
				return stringify(tok, args.get(idx));
			case M_ARG:
				/* Expand the arg. */
				idx = ((Integer) tok.getValue()).intValue();
				// System.out.println("Pushing arg " + args.get(idx));
				arg = args.get(idx).expansion();
				break;
			case M_PASTE:
				paste(tok);
				break;
			default:
				return tok;
			}
		} /* for */
	}

	public String toString() {
		StringBuilder buf = new StringBuilder();
		buf = buf.append("expansion of ").append(macroName);
		Source parent = getParent();
		if (parent != null)
			buf = buf.append(" in ").append(String.valueOf(parent));
		return buf.toString();
	}

	@Override
	String debug_getContent() {
		return macro.getTokens().toString() + " args: " + args;
	}
}

File: PartialPreprocessor/src/main/java/org/anarres/cpp/MacroTokenSource.scala
/*
 * Anarres C Preprocessor
 * Copyright (c) 2007-2008, Shevek
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 */
package org.anarres.cpp

import java.io._
import java.util.Iterator
import java.util.List
import org.anarres.cpp.Token._
import util.control.Breaks
import java.lang.StringBuilder
import collection.JavaConversions.asScalaIterable

object MacroTokenSource {
    /* XXX Called from Preprocessor [ugly]. */
    private[cpp] def escape(buf: StringBuilder, cs: CharSequence): Unit = {
        {
            var i: Int = 0
            while (i < cs.length) {
                {
                    var c: Char = cs.charAt(i)
                    c match {
                        case '\\' =>
                            buf.append("\\\\")
                        case '"' =>
                            buf.append("\\\"")
                        case '\n' =>
                            buf.append("\\n")
                        case '\r' =>
                            buf.append("\\r")
                        case _ =>
                            buf.append(c)
                    }
                }
                ({
                    i += 1; i
                })
            }
        }
    }
}

class MacroTokenSource extends Source {
    private[cpp] def this(macroName: String, m: MacroData, args: List[Argument], gnuCExtensions: Boolean) {
        this()
        this.macroName = macroName
        this.macro = m
        this.tokenIter = m.getTokens.iterator
        this.args = args
        this.arg = null
        this.gnuCExtensions = gnuCExtensions
    }

    private[cpp] override def mayExpand(macroName: String): Boolean = {
        if (macroName.equals(this.macroName)) return false
        return super.mayExpand(macroName)
    }

    private def concat(buf: PrintWriter, arg: Argument, queuedComma: Boolean): Unit = {
        if (queuedComma) {
            if (!arg.isOmittedArg || !gnuCExtensions) {
                buf.append(",")
            }
            else {
                assert(arg.tokens.isEmpty)
                return
            }
        }
        var i: Int = 0
        for (tok <- arg.tokens) {
            if (i != 0 || tok.getType != NL && !tok.isWhite) {
                tok.lazyPrint(buf)
                i += 1
            }
        }
    }

    private def stringify(pos: Token, arg: Argument): Token = {
        var buf: StringWriter = new StringWriter
        var printWriter: PrintWriter = new PrintWriter(buf)
        concat(printWriter, arg, false)
        var str: StringBuilder = new StringBuilder("\"")
        MacroTokenSource.escape(str, buf.getBuffer)
        str.append("\"")
        return new SimpleToken(STRING, pos.getLine, pos.getColumn, str.toString, buf.toString, this)
    }

    private def paste(_ptok: Token): Unit = {
        var ptok = _ptok
        var buf = new StringWriter
        var printWriter = new PrintWriter(buf)
        var queuedComma: Boolean = false
        var count: Int = 2

        var i: Int = 0
        import Breaks._
        breakable {
            while (i < count) {
                if (!tokenIter.hasNext) {
                    error(ptok.getLine, ptok.getColumn, "Paste at end of expansion")
                    buf.append(' ').append(ptok.getText)
                    break
                }
                var tok: Token = tokenIter.next
                if (queuedComma && tok.getType != M_ARG) {
                    buf.append(",")
                    queuedComma = false
                }
                tok.getType match {
                    case M_PASTE =>
                        count += 2
                        ptok = tok
                    case M_ARG =>
                        val idx: Int = (tok.getValue.asInstanceOf[java.lang.Integer]).intValue
                        concat(printWriter, args.get(idx), queuedComma)
                    /* XXX Test this. */
                    case CCOMMENT | CPPCOMMENT=>
                    case ',' =>
                        assert(",".equals(tok.getText))
                        queuedComma = true
                    case _ =>
                        buf.append(tok.getText)
                }
                i += 1
            }
        }
        var sl: StringLexerSource = new StringLexerSource(buf.toString)
        arg = new SourceIterator(sl)
    }

    def token(): Token = {
        val tok: Token = _token()
        if (tok.getType != P_FEATUREEXPR && tok.getText.equals(macroName)) tok.setNoFurtherExpansion
        return tok
    }

    def _token(): Token = {
        while (true) {
            if (arg != null) {
                if (arg.hasNext) {
                    var tok: Token = arg.next
                    assert(tok.getType != M_PASTE, "Unexpected paste token")
                    return tok
                }
                arg = null
            }
            if (!tokenIter.hasNext) return new SimpleToken(EOF, -1, -1, "", this)
            val tok: Token = tokenIter.next
            var idx: Int = 0
            tok.getType match {
                case M_STRING =>
                    idx = (tok.getValue.asInstanceOf[java.lang.Integer]).intValue
                    return stringify(tok, args.get(idx))
                case M_ARG =>
                    idx = (tok.getValue.asInstanceOf[java.lang.Integer]).intValue
                    arg = args.get(idx).expansion
                case M_PASTE =>
                    paste(tok)
                case _ =>
                    return tok
            }
        }
        null
    }

    override def toString: String = {
        var buf: StringBuilder = new StringBuilder
        buf = buf.append("expansion of ").append(macroName)
        var parent: Source = getParent
        if (parent != null) buf = buf.append(" in ").append(String.valueOf(parent))
        return buf.toString
    }

    private[cpp] def debug_getContent: String = {
        return macro.getTokens.toString + " args: " + args
    }

    private final var macro: MacroData = null
    private var tokenIter: Iterator[Token] = null
    private var args: List[Argument] = null
    private var arg: Iterator[Token] = null
    private final var macroName: String = null
    private var gnuCExtensions: Boolean = false
}
